components/provider/src/data_provider.rs

```diff
@@ -2,80 +2,133 @@
 
-// Re-export Error so it can be referenced by "data_provider::Error"
-pub use crate::error::Error;
-
-/// A struct to request a certain hunk of data from a data provider.
-#[derive(PartialEq, Clone, Debug)]
+/// A struct to request a certain piece of data from a data provider.
+#[derive(Clone, Debug, PartialEq)]
 pub struct DataRequest {
-    pub data_key: DataKey,
-    pub data_entry: DataEntry,
+    pub resource_path: ResourcePath,
 }
 
 impl fmt::Display for DataRequest {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}/{}", self.data_key, self.data_entry)
+        write!(
+            f,
+            "{}/{}",
+            self.resource_path.key, self.resource_path.options
+        )
     }
 }
 
-/// A response object containing metadata about the returned data.
-#[derive(Debug, Clone)]
-pub struct DataResponse {
-    pub data_langid: LanguageIdentifier,
+/// Create a DataRequest to a particular ResourceKey with default options.
+impl From<ResourceKey> for DataRequest {
+    fn from(key: ResourceKey) -> Self {
+        DataRequest {
+            resource_path: ResourcePath {
+                key,
+                options: Default::default(),
+            },
+        }
+    }
 }
 
-/// An abstract data provider that loads a payload given a request object.
-pub trait DataProvider<'d> {
-    /// Query the provider for data, loading it into a DataReceiver.
+impl DataRequest {
+    /// Returns the LanguageIdentifier for this DataRequest, or an error if it is not present.
     ///
-    /// Returns Ok if the request successfully loaded data. If data failed to load, returns an
-    /// Error with more information.
-    fn load_to_receiver(
-        &self,
-        req: &DataRequest,
-        receiver: &mut dyn DataReceiver<'d, 'static>,
-    ) -> Result<DataResponse, Error>;
+    /// # Example
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    ///
+    /// const FOO_BAR: ResourceKey = icu_provider::resource_key!(x, "foo", "bar", 1);
+    ///
+    /// let req_no_langid = DataRequest {
+    ///     resource_path: ResourcePath {
+    ///         key: FOO_BAR,
+    ///         options: ResourceOptions::default(),
+    ///     }
+    /// };
+    ///
+    /// let req_with_langid = DataRequest {
+    ///     resource_path: ResourcePath {
+    ///         key: FOO_BAR,
+    ///         options: ResourceOptions {
+    ///             variant: None,
+    ///             langid: Some(icu_locid_macros::langid!("ar-EG")),
+    ///         },
+    ///     }
+    /// };
+    ///
+    /// assert!(matches!(req_no_langid.get_langid(), Err(DataError::NeedsLanguageIdentifier(_))));
+    /// assert!(matches!(req_with_langid.get_langid(), Ok(_)));
+    /// ```
+    pub fn get_langid(&self) -> Result<&LanguageIdentifier, Error> {
+        self.resource_path
+            .options
+            .langid
+            .as_ref()
+            .ok_or_else(|| Error::NeedsLanguageIdentifier(self.clone()))
+    }
+}
```

Comment:
why is it an `Option` on input and `Result` on output? If it's optional to pass, it shouldn't be an error to not have it, so maybe `Option` here as well?

Reply:
I made it `Result` instead of `Option` because it makes it easier to use with the `?` operator.  Otherwise, all the call sites have to call `.ok_or(...)`.

I think of it this way: by calling this function, you are asserting that you require the langid field to be populated in order to handle the request.

Would it be better to name the function something like `try_get_langid()` or `assert_langid()`?

Reply:
If a call site really wants an `Option` instead of `Result`, then they can access `resource_path.options.langid` directly; they don't need the helper function.  I made the helper function specifically because all the call sites had the same boilerplate over and over again of accessing that field and calling `.ok_or` on it.

Reply:
I see. I guess it just looks a bit quirky that you mingle between those two monads which have quite strict designators in Rust.

But I see how in this case your claim is that if someone asks for it, it should be there, and it's an error for it not to be. I would likely rather go with `ok_or`, but I'm not strongly opinionated and I'm ok with it staying as is!

Reply:
LanguageIdentifier is optional only within the context of a specific data request.  If you request plurals data or dates data, then it's required, but if you request unicode property data, then it's ignored.

Would it be better to return `LanguageIdentifier::default()` instead of calling `ok_or`?

Reply:
That would hide the problem. If I understand you correctly, the only scenario in which someone requests a langid is when it must be present, and a lack of it is an error.
Therefore `Result` here seems good. I'd maybe rename the function to `try_langid` - Rust avoids `get_` - https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-[rfc-344] and if it was infallible it'd be `langid()`, so fallible should be `try_langid`.

Reply:
changed to `try_langid`